// @vmw/transport/util/util.js
/*
 * Copyright 2018-2019 VMware, Inc.
 * SPDX-License-Identifier: BSD-2-Clause
 */
export class GeneralUtil {
    static genUUID() {
        let uuid = '', i, random;
        for (i = 0; i < 32; i++) {
            random = Math.random() * 16 | 0;
            if (i === 8 || i === 12 || i === 16 || i === 20) {
                uuid += '-';
            }
            uuid += (i === 12 ? 4 : (i === 16 ? (random & 3 | 8) : random)).toString(16);
        }
        return uuid;
    }
    /**
     * @deprecated do not use, this is not a valid UUID and it confuses our backend systems.
     */
    static genUUIDShort() {
        return GeneralUtil.genUUID().substr(0, 8);
    }
    static isObject(value) {
        try {
            JSON.parse(value);
            return true;
        }
        catch (e) {
            return false;
        }
    }
    /**
     * Returns the value of the specified cookie
     * @param name
     */
    static getCookie(name) {
        const cookieName = name + '=';
        const cookies = document.cookie.split(';');
        for (let i = 0; i < cookies.length; i++) {
            let c = cookies[i];
            while (c.charAt(0) === ' ') {
                c = c.substring(1, c.length);
            }
            if (c.indexOf(cookieName) === 0) {
                return c.substring(cookieName.length, c.length);
            }
        }
        return null;
    }
    /**
     * Return connection string used to establish and manage one or more Fabric connections
     * @param {string} host hostname where Fabric backend is served
     * @param {number} port port where Fabric backend is served
     * @param {string} endpoint target endpoint
     */
    static getFabricConnectionString(host, port, endpoint) {
        if (!host && (port === undefined || port === null)) {
            return `${location.host}${endpoint}`;
        }
        return `${host}:${port}${endpoint}`;
    }
}
