// @vmw/transport/core/services/rest/httpclient.mock.js
/*
 * Copyright 2018 VMware, Inc.
 * SPDX-License-Identifier: BSD-2-Clause
 */
export class MockHttpClient {
    constructor() {
        this.mustFail = false;
        this.errCode = 200;
    }
    delete(uri, requestInit, successHandler, failureHandler) {
        this.httpOperation(uri, requestInit, successHandler, failureHandler);
    }
    get(uri, requestInit, successHandler, failureHandler) {
        this.httpOperation(uri, requestInit, successHandler, failureHandler);
    }
    patch(uri, requestInit, successHandler, failureHandler) {
        this.httpOperation(uri, requestInit, successHandler, failureHandler);
    }
    post(uri, requestInit, successHandler, failureHandler) {
        this.httpOperation(uri, requestInit, successHandler, failureHandler);
    }
    put(uri, requestInit, successHandler, failureHandler) {
        this.httpOperation(uri, requestInit, successHandler, failureHandler);
    }
    httpOperation(uri, requestInit, successHandler, errorHandler) {
        if (this.mustFail) {
            if (this.errCode !== 401) { // to simulate unknown error
                errorHandler(null);
                return;
            }
            const err = {};
            err['status'] = this.errCode;
            err['message'] = 'Fake Error';
            errorHandler(err);
        }
        else {
            successHandler(`${requestInit.method} called`);
        }
    }
}
