// @vmw/transport/bus.api.js
/*
 * Copyright 2017-2020 VMware, Inc.
 * SPDX-License-Identifier: BSD-2-Clause
 */
import { GeneralUtil } from './util/util';
// current version
const version = '1.3.5';
// used by VMW cloud services for store constants.
export const ORG_ID = 'orgId';
export const ORGS = 'orgs';
export var MessageType;
(function (MessageType) {
    MessageType["MessageTypeRequest"] = "Request";
    MessageType["MessageTypeResponse"] = "Response";
    MessageType["MessageTypeError"] = "Error";
    MessageType["MessageTypeControl"] = "Control";
})(MessageType || (MessageType = {}));
export class EventBus {
    /**
     * If you need to reset the ID of this bus, call this, but it may have undesirable effects.
     * This should only be called when re-booting the bus.
     */
    static rebuildId() {
        return `eventbus-${GeneralUtil.genUUID()}-${EventBus.version}`;
    }
}
EventBus.version = version;
EventBus.id = EventBus.rebuildId();
/**
 * Type of transaction? Sync or Async? Default is Async.
 */
export var TransactionType;
(function (TransactionType) {
    TransactionType["ASYNC"] = "Async";
    TransactionType["SYNC"] = "Sync"; // will send requests one at a time and move on after each command is completed in sequence.
})(TransactionType || (TransactionType = {}));
